### PR TITLE
perf(bootstrap): prepare brew bottles concurrently

### PR DESCRIPTION
**Pull request by jdx**
`mise bootstrap packages apply` currently downloads bottles concurrently but then extracts, relocates, signs, and writes every bottle receipt one at a time. This change prepares independent bottles concurrently using the existing `jobs` limit, while preserving dependency-ordered Cellar commits and prefix linking. It does not change the default job count or add a new setting.

For a warm-cache install of 14 bottles from:

```console
mise bootstrap packages apply brew:jq brew:tree brew:wget brew:just brew:shellcheck
```

8 alternating isolated-prefix runs on Apple silicon improved from a 4.097s median on the metadata-only base branch to 2.893s with concurrent staging: 29.4% less time, or 1.42x faster. The comparison reused downloaded bottles and metadata to isolate extraction, relocation, signing, receipt generation, and linking.

Preparation runs in unique, lock-protected staging directories. On preparation failure, queued work is discarded, active work is drained for cleanup, and successfully prepared earlier dependencies still commit in order before the error returns. A later pour reclaims only unlocked staging created with the current lock protocol, leaving lockless directories from older mise processes untouched.

Final Cellar and shared-prefix mutations use a prefix-local commit lock even when operation history is disabled. Bottle preparation remains concurrent outside it; source builds retain it while writing directly into the final Cellar. The installer rechecks completed state under the lock, and creates `opt/<name>` last so a partial public-link operation cannot look complete after a process is killed.

Validation:
- `mise run lint-fix`
- `mise x -- cargo test --locked --bin mise system::packages::brew` (365 passed)
- Installed the benchmark package set into an isolated prefix and ran `jq`, `tree`, `wget`, `just`, and `shellcheck` from that prefix

Stacked on #13151.

*AI-assisted — Tool: Codex; model: unavailable/unavailable; version: unavailable.*

---

> [!NOTE]
> **Medium Risk**
> Changes concurrent filesystem staging, cross-process locking, and the order of prefix linking—core bootstrap install paths—with added tests but real multi-process and partial-failure edge cases.
> 
> **Overview**
> **`mise bootstrap packages apply`** still downloads bottles in parallel, but it now also **extracts, relocates, codesigns, and writes receipts concurrently** (bounded by the existing `jobs` setting). Only the final **Cellar move and prefix linking** stay sequential and dependency-ordered.
> 
> Bottle work is split into **`prepare_bottle`** (lock-protected per-formula staging under the rack) and **`install_prepared`** (shared **`commit_lock`** before mutating Cellar/links). A new **`ConcurrentJobs`** helper can **drop queued prep jobs** after the first failure while **draining in-flight work** so staging locks clean up. Staging adds rack locks, `.mise-lock` files, and reclaim of abandoned staging dirs (without touching legacy lockless dirs). **`opt/<name>` is linked last** so partial linking cannot look “installed,” and source builds **hold the commit lock** for the whole build via **`run_blocking`** with a post-lock **already-installed** recheck.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 015128453493dc576c295aba874c8cf3a72d81f1. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: src/system/packages/brew/mod.rs [hunk 1]
```diff
@@ -263,20 +264,68 @@ impl BrewManager {
...
+                Some(async move {
+                    let result = tokio::task::spawn_blocking(move || {
+                        pour::prepare_bottle(&rf, &tag, &bottle, &tarball, &closure, &*pr)
+                    })
+                    .await
+                    .wrap_err("brew bottle preparation task failed")
+                    .and_then(|result| result);
+                    (index, result)
+                })
+            })
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/system/packages/brew/mod.rs:291]: Addressed in 34d956d04. Bottle preparation now uses a unique `TempDir` inside the formula rack and moves the nested prepared keg into place only during the sequential commit phase. Same-name/version formulae and separate mise processes therefore cannot share staging paths, while the final Cellar/link behavior stays dependency ordered. I reran all 361 Brew-package tests and an isolated-prefix `brew:jq` install.

*AI-assisted — Tool: Codex; model: unavailable/unavailable; version: unavailable.*

File: src/system/packages/brew/pour.rs [hunk 2]
```diff
@@ -252,32 +252,49 @@ pub(super) fn installed_versions(name: &str) -> Vec<String> {
...
-    for dir in [&tmp, &scratch] {
-        if dir.exists() {
-            crate::file::remove_all(dir)?;
-        }
-    }
-    crate::file::create_dir_all(&scratch)?;
+    crate::file::create_dir_all(&rack)?;
+    let staging = tempfile::Builder::new()
+        .prefix(&format!(".mise-extract-{pkg_version}-"))
+        .tempdir_in(&rack)?;
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/system/packages/brew/pour.rs]: Addressed in 133334667. Each staging directory now owns a lock for its full preparation lifetime. A short rack-level lock serializes cleanup with staging creation, so the next pour immediately removes directories whose owners exited or crashed while skipping directories still locked by another active mise process. Legacy fixed `.mise-tmp-*` leftovers are cleaned as well. Added tests for both abandoned cleanup and active-owner preservation; all 363 Brew-package tests pass.

*AI-assisted — Tool: Codex; model: unavailable/unavailable; version: unavailable.*

File: src/system/packages/brew/pour.rs [hunk 3]
```diff
@@ -252,32 +252,91 @@ pub(super) fn installed_versions(name: &str) -> Vec<String> {
...
+            continue;
+        }
+        let Some(lock) = crate::lock_file::LockFile::at(&path.join(".mise-lock")).try_lock()?
+        else {
+            continue;
+        };
+        // The rack lock prevents a new owner from appearing after this check.
+        // Drop the file handle first so cleanup also works on Windows filesystems.
+        drop(lock);
+        crate::file::remove_all(&path)?;
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/system/packages/brew/pour.rs]: Addressed in e04024b79. Cleanup now requires an existing regular `.mise-lock` and opens it directly without creating parent directories, so lockless `.mise-extract-*` and `.mise-tmp-*` directories from older mise processes are preserved. A dedicated regression test covers both legacy names.

*AI-assisted — Tool: Codex; model: unavailable/unavailable; version: unavailable.*

File: src/system/packages/brew/pour.rs [hunk 4]
```diff
@@ -252,32 +252,91 @@ pub(super) fn installed_versions(name: &str) -> Vec<String> {
...
+    }
+    Ok(())
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/system/packages/brew/pour.rs:340]: Addressed in e04024b79. Directory metadata, lock metadata/open, and final removal now treat `NotFound` as a benign concurrent disappearance, so another process dropping its `PreparedBottle` cannot fail cleanup.

*AI-assisted — Tool: Codex; model: unavailable/unavailable; version: unavailable.*

File: src/system/packages/brew/pour.rs [hunk 5]
```diff
@@ -252,32 +252,91 @@ pub(super) fn installed_versions(name: &str) -> Vec<String> {
...
+
+fn remove_abandoned_staging_dirs(rack: &Path) -> Result<()> {
+    for entry in rack.read_dir()? {
+        let entry = entry?;
+        let path = entry.path();
+        let Some(name) = path.file_name().and_then(|name| name.to_str()) else {
+            continue;
+        };
+        if (!name.starts_with(".mise-extract-") && !name.starts_with(".mise-tmp-"))
+            || !path.symlink_metadata()?.is_dir()
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/system/packages/brew/pour.rs]: Addressed in e04024b79. Cleanup skips staging without an existing regular `.mise-lock`; it also avoids the lock helper that creates missing parent directories. Legacy active staging is therefore never inferred to be abandoned.

*AI-assisted — Tool: Codex; model: unavailable/unavailable; version: unavailable.*

File: src/system/packages/brew/pour.rs [hunk 6]
```diff
@@ -310,23 +370,28 @@ pub(super) async fn pour(
...
+/// Callers keep this step sequential and dependency ordered.
+pub(super) fn install_prepared(prepared: PreparedBottle, pr: &dyn SingleReport) -> Result<()> {
     pr.set_message("link".to_string());
-    if keg.exists() {
-        crate::file::remove_all(&keg)?;
+    if prepared.keg.exists() {
+        crate::file::remove_all(&prepared.keg)?;
     }
-    crate::file::rename(&tmp, &keg)?;
+    crate::file::rename(&prepared.staged_keg, &prepared.keg)?;
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/system/packages/brew/pour.rs:440]: Addressed in e04024b79. Bottle commits now hold a prefix-local cross-process lock from final-keg replacement through `link_keg`, independent of operation history. Source builds use the same lock around their direct Cellar build and link path. Download and bottle preparation remain outside it.

*AI-assisted — Tool: Codex; model: unavailable/unavailable; version: unavailable.*

File: src/system/packages/brew/pour.rs [hunk 7]
```diff
@@ -310,23 +419,29 @@ pub(super) async fn pour(
...
 
+/// Commit a prepared bottle into the Cellar and update shared prefix links.
+/// Callers keep this step sequential and dependency ordered.
+pub(super) fn install_prepared(prepared: PreparedBottle, pr: &dyn SingleReport) -> Result<()> {
     pr.set_message("link".to_string());
-    if keg.exists() {
-        crate::file::remove_all(&keg)?;
+    let _commit_lock = commit_lock(pr)?;
+    if prepared.keg.exists() {
+        crate::file::remove_all(&prepared.keg)?;
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/system/packages/brew/pour.rs:438]: Addressed in f346c21c8. Both bottle commits and source builds now recheck `keg_installed` after acquiring the prefix commit lock, so work prepared or downloaded concurrently is discarded instead of replacing an installation another process completed first. The Brew-focused suite passes all 364 tests.

*AI-assisted — Tool: Codex; model: unavailable/unavailable; version: unavailable.*

File: src/system/packages/brew/source.rs [hunk 8]
```diff
@@ -125,6 +125,7 @@ pub(super) async fn build(
 
     // formulae bake the final keg path into binaries, so the build installs
     // straight into the Cellar (same as brew); a failed build removes the keg
+    let _commit_lock = pour::commit_lock(pr)?;
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/system/packages/brew/source.rs]: Addressed in f346c21c8. Source builds acquire the commit lock before touching the shared build root and retain it through linking. The Ruby build now uses the existing cancellation-safe synchronous runner inside `run_blocking`, so the future cannot release the lock while a child continues writing to Cellar, without broad changes to generic async command execution. Bottle downloads and preparation remain outside the lock. The Brew-focused suite passes all 364 tests and `mise run lint-fix` passes.

*AI-assisted — Tool: Codex; model: unavailable/unavailable; version: unavailable.*

File: src/system/packages/brew/pour.rs [hunk 9]
```diff
@@ -310,23 +420,34 @@ pub(super) async fn pour(
...
+pub(super) fn install_prepared(prepared: PreparedBottle, pr: &dyn SingleReport) -> Result<()> {
     pr.set_message("link".to_string());
-    if keg.exists() {
-        crate::file::remove_all(&keg)?;
+    let _commit_lock = commit_lock(pr)?;
+    // Another mise process may have completed the same formula while this
+    // process prepared its bottle outside the commit lock.
+    if keg_installed(&prepared.name, &prepared.pkg_version) {
+        return Ok(());
+    }
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/system/packages/brew/pour.rs:436]: Addressed in b4e6fc9d0. `link_keg` now creates all public links and the linked-keg record before creating `opt/<name>` last. Because `keg_installed` uses `opt` as its completion marker, a process killed during linking leaves no completed marker; a waiter will replace and fully relink that partial keg instead of returning success. The Brew-focused suite passes all 364 tests.

*AI-assisted — Tool: Codex; model: unavailable/unavailable; version: unavailable.*

File: src/system/packages/brew/mod.rs [hunk 10]
```diff
@@ -263,20 +264,68 @@ impl BrewManager {
...
+                    for (other_index, pending) in reports.iter().enumerate() {
+                        if other_index != index {
+                            pending.abandon();
+                        }
+                    }
+                    mpr.footer_finish();
+                    return Err(err);
+                }
+            }
+        }
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/system/packages/brew/mod.rs:305]: Addressed in b4e6fc9d0. Preparation now uses a bounded scheduler with separate pending and active sets. The first preparation error drops queued work, drains active jobs for safe staging cleanup, and retains their indexed results. Dependency-ordered commit then installs successful earlier formulae before returning the first preparation failure, matching the previous partial-progress behavior without preparing every remaining bottle. Added a scheduler regression test; all 364 Brew-package tests and `mise run lint-fix` pass.

*AI-assisted — Tool: Codex; model: unavailable/unavailable; version: unavailable.*

File: src/system/packages/brew/pour.rs [hunk 11]
```diff
@@ -655,6 +775,10 @@ pub(super) fn link_keg(name: &str, pkg_version: &str, keg_only: bool) -> Result<
             conflicts.push(linked);
         }
     }
+    // Create opt last: keg_installed uses it as the completion marker, so a
+    // process killed during public linking cannot make a partial pour look
+    // complete to another process waiting on the commit lock.
+    links.push((opt_link.clone(), keg.clone()));
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/system/packages/brew/pour.rs]: Fixed. The opt link is still appended last so it remains the installation-completion marker, but it now passes through can_overwrite during the same preflight conflict scan as every other public link. Added a regression test that places a foreign regular file at opt/foo, verifies linking fails before any public links are created, and verifies the file remains unchanged. The 365-test Brew suite and full lint/fix pipeline pass.

*AI-assisted — Tool: Codex; model: unavailable/unavailable; version: unavailable.*
